bookshelf/src/DataModel/Datastore.php

```diff
@@ -0,0 +1,246 @@
+            'limit' => $limit,
+            'startCursor' => $cursor,
+        ]);
+
+        $request = new \Google_Service_Datastore_RunQueryRequest();
+        $request->setQuery($query);
+        $response = $this->datastore->datasets->
+            runQuery($this->datasetId, $request);
+
+        /** @var \Google_Service_Datastore_QueryResultBatch $batch */
+        $batch = $response->getBatch();
+        $endCursor = $batch->getEndCursor();
+
+        return array(
+            'books' => $response,
+            'cursor' => $endCursor === $cursor ? null : $endCursor,
+        );
+    }
+
+    public function create($book, $key = null)
+    {
+        $this->verifyBook($book);
+
+        if (is_null($key)) {
+            $key = $this->createKey();
+        }
+
+        $properties = $this->bookToProperties($book);
+
+        $entity = new \Google_Service_Datastore_Entity([
+            'key' => $key,
+            'properties' => $properties
+        ]);
+
+        $request = new \Google_Service_Datastore_CommitRequest([
+            'mode' => 'NON_TRANSACTIONAL',
+            'mutation' => [
+                'upsert' => [$entity]
+            ]
+        ]);
+
+        $this->datastore->datasets->commit($this->datasetId, $request);
+
+        // return the ID of the created datastore item
+        return $key->getPath()[0]->getId();
+    }
+
+    public function read($id)
+    {
+        $key = $this->createKey($id);
+        $request = new \Google_Service_Datastore_LookupRequest([
+            'keys' => [$key]
+        ]);
+
+        $response = $this->datastore->datasets->
+            lookup($this->datasetId, $request);
+
+        /** @var \Google_Service_Datastore_QueryResultBatch $batch */
+        if ($found = $response->getFound()) {
+            $book = $this->propertiesToBook($found[0]['entity']['properties']);
+            $book['id'] = $id;
+
+            return $book;
+        }
+
+        return false;
+    }
+
+    public function update($book)
+    {
+        $this->verifyBook($book);
+
+        if (!isset($book['id'])) {
+            throw new InvalidArgumentException('Book must have an "id" attribute');
+        }
+
+        $key = $this->createKey($book['id']);
+        $properties = $this->bookToProperties($book);
+
+        $entity = new \Google_Service_Datastore_Entity([
+            'key' => $key,
+            'properties' => $properties
+        ]);
+
+        $request = new \Google_Service_Datastore_CommitRequest([
+            'mode' => 'NON_TRANSACTIONAL',
+            'mutation' => [
+                'update' => [$entity]
+            ]
+        ]);
+
+        $response = $this->datastore->datasets->commit($this->datasetId, $request);
+
+        // return the number of updated rows
+        return 1;
+    }
+
+    public function delete($id)
+    {
+        $key = $this->createKey($id);
+
+        $request = new \Google_Service_Datastore_CommitRequest([
+            'mode' => 'NON_TRANSACTIONAL',
```

Comment:
Ahh, here's the reason. If you want to use transactional commits, you have to wrap everything in a transaction request, which is more complicated and not really necessary for this example (we're only making one request at a time, there's no benefit in wrapping it in a transaction). Otherwise, you get this error:

```json
"domain": "global",
"reason": "INVALID_ARGUMENT",
"message": "transactional commit requires a transaction"
```

So I think we should keep it as-is